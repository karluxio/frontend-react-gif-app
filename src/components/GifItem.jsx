import PropTypes from "prop-types"

export const GifItem = ({ id, title, url }) => {
  return (

    <div className="card">
      <img src={url} />
      <p>{title}</p>
    </div>

  )
}

GifItem.prototypes = {
  id: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
}
