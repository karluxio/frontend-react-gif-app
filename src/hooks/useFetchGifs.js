import { useEffect, useState } from "react"
import { fetchGifs } from "../helpers/fetchGifs"

export const useFetchGifs = (category) => {

  const [data, setData] = useState([])
  const [loading, setLoading] = useState(true)
  const [error, setError] = useState(true)

  const getGifs = async () => {
    const gifsFromApi = await fetchGifs(category)

    setData(gifsFromApi)
    setLoading(false)
    setError(false)
  }

  useEffect(() => {
    getGifs()
  }, [])

  return {
    data, loading, error
  }
}
