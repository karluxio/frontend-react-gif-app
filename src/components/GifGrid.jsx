import { useFetchGifs } from "../hooks/useFetchGifs";
import { GifItem } from "./GifItem";

export const GifGrid = ({ category }) => {

  const { data, loading, error } = useFetchGifs(category);

  console.log(data);

  return (
    <>
      <h3>{category}</h3>

      {loading && <h3>Loading...</h3>}

      {error && <h3>Error...</h3>}

      <div className="card-grid">
        {!loading && !error && data.map(({ id, title, images }) => (
          <GifItem
            key={id}
            id={id}
            title={title}
            url={images.downsized_medium.url}
          />
        ))}
      </div>

    </>
  )
}
