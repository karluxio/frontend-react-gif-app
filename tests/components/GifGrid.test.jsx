import { render, screen } from "@testing-library/react"
import { GifGrid } from "../../src/components/GifGrid"
import { useFetchGifs } from "../../src/hooks/useFetchGifs"

jest.mock("../../src/hooks/useFetchGifs")

describe('testing on <GifGrid />', () => {
  const category = 'Goku'

  test("should match against snapshot", () => {

    useFetchGifs.mockReturnValue({
      images: [],
      isLoading: true
    });

    const { container } = render(<GifGrid category={category} />)

    expect(container).toMatchSnapshot()
  })

  test("should shows loading initially", () => {

    useFetchGifs.mockReturnValue({
      images: [],
      isLoading: true
    });

    render(<GifGrid category={category} />);
    expect(screen.getByText(/loading/i));
    expect(screen.getByText(category));
  })

  test('should shows items when gifs are loading from Api', () => {
    const gifs = [
      {
        id: 'ABC',
        title: 'Saitama',
        url: 'https://localhost/saitama.jpg'
      },
      {
        id: '123',
        title: 'Goku',
        url: 'https://localhost/goku.jpg'
      }
    ]

    useFetchGifs.mockReturnValue({
      images: gifs,
      isLoading: false
    });

    render(<GifGrid category={category} />)
    expect(screen.getAllByRole('img').length).toBe(2);

  })
})