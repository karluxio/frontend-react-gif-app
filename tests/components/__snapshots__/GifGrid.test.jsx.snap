// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing on <GifGrid /> should match against snapshot 1`] = `
<div>
  <h3>
    Goku
  </h3>
  <h3>
    Loading...
  </h3>
  <div
    class="card-grid"
  />
</div>
`;
